rfd/0194-upstream-pki-support.md

```diff
@@ -0,0 +1,146 @@
+Teleport Workload Identity credentials are issued by an intermediate to a 
+root certificate authority that is controlled by the organization.
+
+For example:
+
+- It may be required as part of compliance policy. 
+- Ability to revoke certificates issued by Teleport Workload Identity, without
+  using Teleport Workload Identity.
+- Ability to rotate the Teleport Workload Identity certificate authority without
+  needing to distribute the new certificate authority to validators working with
+  workload identities.
+
+## Details
+
+The project is largely divided into two phases:
+
+- Building out the underlying support for Teleport CAs being intermediate CAs
+  within an external PKI hierarchy, with a manual process for rotating.
+- Building out integrations with specific PKI solutions to automate the process
+  of rotating the Teleport CA.
+
+### Phase 1: Intermediate CA Support
+
+#### Overview
+
+A new `pre-init` phase will be added to the Teleport CA rotation process.
+
+The pre-init phase may only be entered from the `standby` phase.
+
+Upon entering the `pre-init` phase, a new set of key pairs
+(but not certificates) will be generated and stored in the
+CertificateAuthorityV2 resource. These will be stored in a new field,
+`spec.NextKeys` of the existing CAKeySet type.
+
+If the user requests that the CA return to the `standby` phase from `pre-init`,
+then the `NextKeys` field will be cleared. No further action is required.
+
+Once in the `pre-init` phase, the user may generate a CSR for the new CA key-pair
+using the `tctl auth generate-csr` command.
+
+To progress from the `pre-init` phase to the `init` phase, the user must
+provide a signed certificate for the new CA key-pair, as well as the chain of
+intermediate CAs to and including the root CA.
+
+The following process will then occur:
+
+- Teleport will verify the signed certificate and chain:
+  - The given leaf CA validates against the given chain.
+  - The given leaf CA's key matches the generated key in the `NextKeys` field.
+- The CertificateAuthority resource will be atomically updated:
+  - The given leaf CA will be written into `spec.NextKeys.TLS.Cert`.
+  - The given intermediate CAs will be written into `spec.NextKeys.TLS.IntermediateCerts`.
+  - The given root CAs will be written into `spec.NextKeys.TLS.RootCerts`.
+  - The `NextKeys` field will be swapped into the `AdditionalTrustedKeys` field. 
+  - To enter the `init` phase.
+
+Once in the `init` phase, the rotation process will proceed as normal.
+
+#### Resource Changes
+
+The following new fields will be added to the `CertificateAuthorityV2` protobuf
+message:
+
+- `NextKeys` (CAKeySet): The keys generated upon entering the `pre-init` phase
+  and for which a CSR can be generated.
+
+To accommodate storing the intermediate and root CAs related to a Teleport
+CA, two new fields will be introduced to the `TLSKeyPair` protobuf message:
+
+- `IntermediateCerts` (bytes) - PEM-encoded intermediate certs linking the
```

Comment:
@nklaassen - I figured that storing the intermediate certs in a new field rather than chaining these in the existing `Cert` field was less likely to break any existing code that expects `Cert` to contain a single X509 certificate. WDYT?

Reply:
Probably good to keep the root and intermediates separate from cert, but now I'm wondering is there a reason `IntermediateCerts` and `RootCerts` are separate fields and not one single chain?